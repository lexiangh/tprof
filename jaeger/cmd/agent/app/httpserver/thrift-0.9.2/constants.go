// Copyright (c) 2019 The Jaeger Authors.
// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Autogenerated by Thrift Compiler (0.9.2)
// DO NOT EDIT UNLESS YOU ARE SURE THAT YOU KNOW WHAT YOU ARE DOING

package sampling

import (
	"bytes"
	"fmt"

	"github.com/apache/thrift/lib/go/thrift"
)

// (needed to ensure safety because of naive import list construction.)
var _ = thrift.ZERO
var _ = fmt.Printf
var _ = bytes.Equal

func init() {
}
