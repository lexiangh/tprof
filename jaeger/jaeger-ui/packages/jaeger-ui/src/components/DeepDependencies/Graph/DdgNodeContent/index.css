/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgNodeContent--core {
  background: #eee;
  border-radius: 100%;
  box-shadow: 0 1px 2px 1px rgba(0, 0, 0, 0.15);
  cursor: default;
  display: flex;
}

.DdgNodeContent--core.is-positioned {
  bottom: 1px;
  left: 1px;
  position: absolute;
  right: 1px;
  top: 1px;
}

.DdgNodeContent--labelWrapper {
  margin-right: auto;
  margin-left: auto;
}

.DdgNodeContent--label {
  margin: 0;
  margin-right: auto;
  margin-left: auto;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.DdgNodeContent--actionsWrapper {
  background: #fff;
  border-radius: 3px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  bottom: 100%;
  box-shadow: 0 0px 4px 1px rgba(0, 0, 0, 0.1);
  left: 1em;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  transition: opacity 0.1s;
  white-space: nowrap;
}

.DdgNodeContent:hover > .DdgNodeContent--actionsWrapper {
  opacity: 1;
  pointer-events: all;
}

.DdgNodeContent--actionsItem {
  align-items: center;
  display: flex;
  padding: 0.5em;
}

.DdgNodeContent--actionsItem:not(:hover) {
  color: inherit;
}

.DdgNodeContent--actionsItemText {
  font-size: 0.9em;
  line-height: normal;
  margin-left: 0.5em;
}
