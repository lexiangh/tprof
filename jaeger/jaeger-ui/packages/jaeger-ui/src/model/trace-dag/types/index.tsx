// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Span } from '../../../types/trace';
import { TNil } from '../../../types';

export type NodeID = string;

export type DiffCounts = {
  a: number;
  b: number;
};

export type DenseSpan = {
  span: Span;
  id: string;
  service: string;
  operation: string;
  tags: Record<string, any>;
  parentID: string | TNil;
  skipToChild: boolean;
  children: Set<string>;
};
