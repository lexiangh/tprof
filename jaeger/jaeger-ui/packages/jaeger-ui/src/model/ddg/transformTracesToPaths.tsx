// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import memoizeOne from 'memoize-one';

import spanAncestorIds from '../../utils/span-ancestor-ids';

import { TDdgPayloadEntry, TDdgPayloadPath, TDdgPayload } from './types';
import { FetchedTrace } from '../../types';
import { Span, Trace } from '../../types/trace';

function convertSpan(span: Span, trace: Trace): TDdgPayloadEntry {
  const serviceName = trace.processes[span.processID].serviceName;
  const operationName = span.operationName;
  return { service: serviceName, operation: operationName };
}

function transformTracesToPaths(
  traces: Record<string, FetchedTrace>,
  focalService: string,
  focalOperation: string | undefined
): TDdgPayload {
  const dependencies: TDdgPayloadPath[] = [];
  Object.values(traces).forEach(({ data }) => {
    if (data) {
      const spanMap: Map<string, Span> = new Map();
      const { traceID } = data;
      data.spans
        .filter(span => {
          spanMap.set(span.spanID, span);
          return !span.hasChildren;
        })
        .forEach(leaf => {
          const path = spanAncestorIds(leaf).map(id => {
            const span = spanMap.get(id);
            if (!span) throw new Error(`Ancestor spanID ${id} not found in trace ${traceID}`);

            return convertSpan(span, data);
          });
          path.push(convertSpan(leaf, data));

          if (
            path.some(
              ({ service, operation }) =>
                service === focalService && (!focalOperation || operation === focalOperation)
            )
          ) {
            dependencies.push({
              path,
              attributes: [
                {
                  key: 'exemplar_trace_id',
                  value: traceID,
                },
              ],
            });
          }
        });
    }
  });
  return { dependencies };
}

export default memoizeOne(transformTracesToPaths);
