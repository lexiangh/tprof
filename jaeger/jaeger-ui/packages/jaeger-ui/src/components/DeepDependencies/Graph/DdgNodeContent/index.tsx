// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';

import calcPositioning from './calc-positioning';
import {
  MAX_LENGTH,
  MAX_LINKED_TRACES,
  MIN_LENGTH,
  OP_PADDING_TOP,
  PARAM_NAME_LENGTH,
  RADIUS,
  WORD_RX,
} from './constants';
import { setFocusIcon } from './node-icons';
import { getUrl } from '../../url';
import BreakableText from '../../../common/BreakableText';
import NewWindowIcon from '../../../common/NewWindowIcon';
import { getUrl as getSearchUrl } from '../../../SearchTracePage/url';
import { EDdgDensity, EViewModifier, TDdgVertex, PathElem } from '../../../../model/ddg/types';

import './index.css';

type TProps = {
  focalNodeUrl: string | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  isFocalNode: boolean;
  isPositioned: boolean;
  operation: string | null;
  service: string;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, isEnabled: boolean) => void;
  vertexKey: string;
};

export default class DdgNodeContent extends React.PureComponent<TProps> {
  static measureNode() {
    const diameter = 2 * (RADIUS + 1);

    return {
      height: diameter,
      width: diameter,
    };
  }

  static getNodeRenderer(
    getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined,
    setViewModifier: (vertexKey: string, viewModifier: EViewModifier, enable: boolean) => void,
    density: EDdgDensity,
    showOp: boolean,
    baseUrl: string,
    extraUrlArgs: { [key: string]: unknown } | undefined
  ) {
    return function renderNode(vertex: TDdgVertex, _: unknown, lv: TLayoutVertex<any> | null) {
      const { isFocalNode, key, operation, service } = vertex;
      return (
        <DdgNodeContent
          focalNodeUrl={
            isFocalNode ? null : getUrl({ density, operation, service, showOp, ...extraUrlArgs }, baseUrl)
          }
          getVisiblePathElems={getVisiblePathElems}
          isFocalNode={isFocalNode}
          isPositioned={Boolean(lv)}
          operation={operation}
          setViewModifier={setViewModifier}
          service={service}
          vertexKey={key}
        />
      );
    };
  }

  private viewTraces = () => {
    const { vertexKey, getVisiblePathElems } = this.props;
    const elems = getVisiblePathElems(vertexKey);
    if (elems) {
      const urlIds: Set<string> = new Set();
      let currLength = MIN_LENGTH;
      // Because there is a limit on traceIDs, attempt to get some from each elem rather than all from one.
      const allIDs = elems.map(({ memberOf }) => memberOf.traceIDs.slice());
      while (allIDs.length) {
        const ids = allIDs.shift();
        if (ids && ids.length) {
          const id = ids.pop();
          if (id && !urlIds.has(id)) {
            // Keep track of the length, then break if it is too long, to avoid opening a tab with a URL that
            // the backend cannot process, even if there are more traceIDs
            currLength += PARAM_NAME_LENGTH + id.length;
            if (currLength > MAX_LENGTH) {
              break;
            }
            urlIds.add(id);
            if (urlIds.size >= MAX_LINKED_TRACES) {
              break;
            }
          }
          allIDs.push(ids);
        }
      }
      window.open(getSearchUrl({ traceID: Array.from(urlIds) }), '_blank');
    }
  };

  private onMouseUx = (event: React.MouseEvent<HTMLElement>) => {
    const { vertexKey, setViewModifier } = this.props;
    setViewModifier(vertexKey, EViewModifier.Hovered, event.type === 'mouseover');
  };

  render() {
    const { focalNodeUrl, isFocalNode, isPositioned, operation, service } = this.props;
    const { radius, svcWidth, opWidth, svcMarginTop } = calcPositioning(service, operation);
    const scaleFactor = RADIUS / radius;
    const transform = `translate(${RADIUS - radius}px, ${RADIUS - radius}px) scale(${scaleFactor})`;
    return (
      <div className="DdgNodeContent" onMouseOver={this.onMouseUx} onMouseOut={this.onMouseUx}>
        <div
          className={cx('DdgNodeContent--core', {
            'is-focalNode': isFocalNode,
            'is-positioned': isPositioned,
          })}
          style={{ width: `${radius * 2}px`, height: `${radius * 2}px`, transform }}
        >
          <div className="DdgNodeContent--labelWrapper">
            <h4
              className="DdgNodeContent--label"
              style={{ marginTop: `${svcMarginTop}px`, width: `${svcWidth}px` }}
            >
              <BreakableText text={service} wordRegexp={WORD_RX} />
            </h4>
            {operation && (
              <div
                className="DdgNodeContent--label"
                style={{ paddingTop: `${OP_PADDING_TOP}px`, width: `${opWidth}px` }}
              >
                <BreakableText text={operation} wordRegexp={WORD_RX} />
              </div>
            )}
          </div>
        </div>

        <div className="DdgNodeContent--actionsWrapper">
          {focalNodeUrl && (
            <a href={focalNodeUrl} className="DdgNodeContent--actionsItem">
              {setFocusIcon}
              <span className="DdgNodeContent--actionsItemText">Set focus</span>
            </a>
          )}
          <a className="DdgNodeContent--actionsItem" onClick={this.viewTraces} role="button">
            <NewWindowIcon />
            <span className="DdgNodeContent--actionsItemText">View traces</span>
          </a>
        </div>
      </div>
    );
  }
}
