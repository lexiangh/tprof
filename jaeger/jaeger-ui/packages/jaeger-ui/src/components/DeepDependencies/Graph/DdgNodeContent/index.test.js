// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

jest.mock('./calc-positioning', () => () => ({
  radius: 50,
  svcWidth: 20,
  opWidth: 30,
  svcMarginTop: 10,
}));

/* eslint-disable import/first */
import React from 'react';
import { shallow } from 'enzyme';

import DdgNodeContent from '.';
import { MAX_LENGTH, MAX_LINKED_TRACES, MIN_LENGTH, PARAM_NAME_LENGTH, RADIUS } from './constants';
import * as getSearchUrl from '../../../SearchTracePage/url';

import { EDdgDensity, EViewModifier } from '../../../../model/ddg/types';

describe('<DdgNodeContent>', () => {
  const vertexKey = 'some-key';
  const service = 'some-service';
  const operation = 'some-operation';
  const props = {
    focalNodeUrl: 'some-url',
    getVisiblePathElems: jest.fn(),
    isFocalNode: false,
    operation,
    setViewModifier: jest.fn(),
    service,
    vertexKey,
  };

  let wrapper;

  beforeEach(() => {
    props.getVisiblePathElems.mockReset();
    wrapper = shallow(<DdgNodeContent {...props} />);
  });

  it('does not explode', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('omits the operation if it is null', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ operation: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('calls setViewModifier on mouse over, out', () => {
    const { calls } = props.setViewModifier.mock;
    wrapper.simulate('mouseover', { type: 'mouseover' });
    expect(calls.length).toBe(1);
    wrapper.simulate('mouseout', { type: 'mouseout' });
    expect(calls.length).toBe(2);
    expect(calls[0]).toEqual([vertexKey, EViewModifier.Hovered, true]);
    expect(calls[1]).toEqual([vertexKey, EViewModifier.Hovered, false]);
  });

  it('renders correctly when isFocalNode = true and focalNodeUrl = null', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ focalNodeUrl: null, isFocalNode: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('measureNode', () => {
    it('returns twice the RADIUS with a buffer for svg border', () => {
      const diameterWithBuffer = 2 * RADIUS + 2;
      expect(DdgNodeContent.measureNode()).toEqual({
        height: diameterWithBuffer,
        width: diameterWithBuffer,
      });
    });
  });

  describe('viewTraces', () => {
    const click = () =>
      wrapper
        .find('.DdgNodeContent--actionsItem')
        .at(1)
        .simulate('click');
    const pad = num => `000${num}`.slice(-4);
    const mockReturn = ids =>
      props.getVisiblePathElems.mockReturnValue(ids.map(traceIDs => ({ memberOf: { traceIDs } })));
    const calcIdxWithinLimit = arr => Math.floor(0.75 * arr.length);
    const falsifyDuplicateAndMock = ids => {
      const withFalsyAndDuplicate = ids.map(arr => arr.slice());
      withFalsyAndDuplicate[0].splice(
        calcIdxWithinLimit(withFalsyAndDuplicate[0]),
        0,
        withFalsyAndDuplicate[1][calcIdxWithinLimit(withFalsyAndDuplicate[1])],
        ''
      );
      withFalsyAndDuplicate[1].splice(
        calcIdxWithinLimit(withFalsyAndDuplicate[1]),
        0,
        withFalsyAndDuplicate[0][calcIdxWithinLimit(withFalsyAndDuplicate[0])],
        ''
      );
      mockReturn(withFalsyAndDuplicate);
    };
    const makeIDsAndMock = (idCounts, makeID = count => `test traceID${count}`) => {
      let idCount = 0;
      const ids = idCounts.map(count => {
        const rv = [];
        for (let i = 0; i < count; i++) {
          rv.push(makeID(pad(idCount++)));
        }
        return rv;
      });
      mockReturn(ids);
      return ids;
    };
    let getSearchUrlSpy;
    const lastIDs = () => getSearchUrlSpy.mock.calls[getSearchUrlSpy.mock.calls.length - 1][0].traceID;
    let originalOpen;

    beforeAll(() => {
      originalOpen = window.open;
      window.open = jest.fn();
      getSearchUrlSpy = jest.spyOn(getSearchUrl, 'getUrl');
    });

    beforeEach(() => {
      window.open.mockReset();
    });

    afterAll(() => {
      window.open = originalOpen;
    });

    it('no-ops if there are no elems for key', () => {
      props.getVisiblePathElems.mockReturnValue();
      click();
      expect(window.open).not.toHaveBeenCalled();
    });

    it('opens new tab viewing single traceID from single elem', () => {
      const ids = makeIDsAndMock([1]);
      click();

      expect(lastIDs().sort()).toEqual([].concat(...ids).sort());
      expect(props.getVisiblePathElems).toHaveBeenCalledTimes(1);
      expect(props.getVisiblePathElems).toHaveBeenCalledWith(vertexKey);
    });

    it('opens new tab viewing multiple traceIDs from single elem', () => {
      const ids = makeIDsAndMock([3]);
      click();

      expect(lastIDs().sort()).toEqual([].concat(...ids).sort());
    });

    it('opens new tab viewing multiple traceIDs from multiple elems', () => {
      const ids = makeIDsAndMock([3, 2]);
      click();

      expect(lastIDs().sort()).toEqual([].concat(...ids).sort());
    });

    it('ignores falsy and duplicate IDs', () => {
      const ids = makeIDsAndMock([3, 3]);
      falsifyDuplicateAndMock(ids);
      click();

      expect(lastIDs().sort()).toEqual([].concat(...ids).sort());
    });

    describe('MAX_LINKED_TRACES', () => {
      const ids = makeIDsAndMock([MAX_LINKED_TRACES, MAX_LINKED_TRACES, 1]);
      const expected = [
        ...ids[0].slice(MAX_LINKED_TRACES / 2 + 1),
        ...ids[1].slice(MAX_LINKED_TRACES / 2 + 1),
        ids[2][0],
      ].sort();

      it('limits link to only include MAX_LINKED_TRACES, taking equal from each pathElem', () => {
        mockReturn(ids);
        click();

        expect(lastIDs().sort()).toEqual(expected);
      });

      it('does not count falsy and duplicate IDs towards MAX_LINKED_TRACES', () => {
        falsifyDuplicateAndMock(ids);
        click();

        expect(lastIDs().sort()).toEqual(expected);
      });
    });

    describe('MAX_LENGTH', () => {
      const effectiveMaxLength = MAX_LENGTH - MIN_LENGTH;
      const TARGET_ID_COUNT = 31;
      const paddingLength = Math.floor(effectiveMaxLength / TARGET_ID_COUNT) - PARAM_NAME_LENGTH;
      const idPadding = 'x'.repeat(paddingLength - pad(0).length);
      const ids = makeIDsAndMock([TARGET_ID_COUNT, TARGET_ID_COUNT, 1], num => `${idPadding}${num}`);
      const expected = [
        ...ids[0].slice(TARGET_ID_COUNT / 2 + 1),
        ...ids[1].slice(TARGET_ID_COUNT / 2 + 1),
        ids[2][0],
      ].sort();

      it('limits link to only include MAX_LENGTH, taking equal from each pathElem', () => {
        mockReturn(ids);
        click();

        expect(lastIDs().sort()).toEqual(expected);
      });

      it('does not count falsy and duplicate IDs towards MAX_LEN', () => {
        falsifyDuplicateAndMock(ids);
        click();

        expect(lastIDs().sort()).toEqual(expected);
      });
    });
  });

  describe('DdgNodeContent.getNodeRenderer()', () => {
    const ddgVertex = {
      isFocalNode: false,
      key: 'some-key',
      operation: 'the-operation',
      service: 'the-service',
    };
    const noOp = () => {};

    it('returns a <DdgNodeContent />', () => {
      const ddgNode = DdgNodeContent.getNodeRenderer(
        noOp,
        noOp,
        EDdgDensity.PreventPathEntanglement,
        true,
        'testBaseUrl',
        { maxDuration: '100ms' }
      )(ddgVertex);
      expect(ddgNode).toBeDefined();
      expect(shallow(ddgNode)).toMatchSnapshot();
      expect(ddgNode.type).toBe(DdgNodeContent);
    });

    it('returns a focal <DdgNodeContent />', () => {
      const focalNode = DdgNodeContent.getNodeRenderer(noOp, noOp)({
        ...ddgVertex,
        isFocalNode: true,
      });
      expect(focalNode).toBeDefined();
      expect(shallow(focalNode)).toMatchSnapshot();
      expect(focalNode.type).toBe(DdgNodeContent);
    });
  });
});
