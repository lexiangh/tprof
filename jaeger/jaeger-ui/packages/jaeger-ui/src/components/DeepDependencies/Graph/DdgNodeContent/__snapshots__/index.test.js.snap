// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DdgNodeContent> DdgNodeContent.getNodeRenderer() returns a <DdgNodeContent /> 1`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="the-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
      <div
        className="DdgNodeContent--label"
        style={
          Object {
            "paddingTop": "5px",
            "width": "30px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="the-operation"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </div>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      href="testBaseUrl?density=ppe&maxDuration=100ms&operation=the-operation&service=the-service&showOp=1"
    >
      <svg
        className="DdgNode--SetFocusIcon"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M50.0001 -1L61.0557 22.1383H38.9444L50.0001 -1Z"
            fill="currentColor"
          />
          <path
            d="M49.9999 99L38.9443 75.8617L61.0556 75.8617L49.9999 99Z"
            fill="currentColor"
          />
          <path
            d="M100 49L76.8617 60.0556L76.8617 37.9444L100 49Z"
            fill="currentColor"
          />
          <path
            d="M1.57361e-06 49L23.1383 37.9444L23.1383 60.0556L1.57361e-06 49Z"
            fill="currentColor"
          />
        </g>
      </svg>
      <span
        className="DdgNodeContent--actionsItemText"
      >
        Set focus
      </span>
    </a>
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;

exports[`<DdgNodeContent> DdgNodeContent.getNodeRenderer() returns a focal <DdgNodeContent /> 1`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core is-focalNode"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="the-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
      <div
        className="DdgNodeContent--label"
        style={
          Object {
            "paddingTop": "5px",
            "width": "30px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="the-operation"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </div>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;

exports[`<DdgNodeContent> omits the operation if it is null 1`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
      <div
        className="DdgNodeContent--label"
        style={
          Object {
            "paddingTop": "5px",
            "width": "30px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-operation"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </div>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      href="some-url"
    >
      <svg
        className="DdgNode--SetFocusIcon"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M50.0001 -1L61.0557 22.1383H38.9444L50.0001 -1Z"
            fill="currentColor"
          />
          <path
            d="M49.9999 99L38.9443 75.8617L61.0556 75.8617L49.9999 99Z"
            fill="currentColor"
          />
          <path
            d="M100 49L76.8617 60.0556L76.8617 37.9444L100 49Z"
            fill="currentColor"
          />
          <path
            d="M1.57361e-06 49L23.1383 37.9444L23.1383 60.0556L1.57361e-06 49Z"
            fill="currentColor"
          />
        </g>
      </svg>
      <span
        className="DdgNodeContent--actionsItemText"
      >
        Set focus
      </span>
    </a>
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;

exports[`<DdgNodeContent> omits the operation if it is null 2`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      href="some-url"
    >
      <svg
        className="DdgNode--SetFocusIcon"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M50.0001 -1L61.0557 22.1383H38.9444L50.0001 -1Z"
            fill="currentColor"
          />
          <path
            d="M49.9999 99L38.9443 75.8617L61.0556 75.8617L49.9999 99Z"
            fill="currentColor"
          />
          <path
            d="M100 49L76.8617 60.0556L76.8617 37.9444L100 49Z"
            fill="currentColor"
          />
          <path
            d="M1.57361e-06 49L23.1383 37.9444L23.1383 60.0556L1.57361e-06 49Z"
            fill="currentColor"
          />
        </g>
      </svg>
      <span
        className="DdgNodeContent--actionsItemText"
      >
        Set focus
      </span>
    </a>
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;

exports[`<DdgNodeContent> renders correctly when isFocalNode = true and focalNodeUrl = null 1`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
      <div
        className="DdgNodeContent--label"
        style={
          Object {
            "paddingTop": "5px",
            "width": "30px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-operation"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </div>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      href="some-url"
    >
      <svg
        className="DdgNode--SetFocusIcon"
        height="100"
        viewBox="0 0 100 100"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M50.0001 -1L61.0557 22.1383H38.9444L50.0001 -1Z"
            fill="currentColor"
          />
          <path
            d="M49.9999 99L38.9443 75.8617L61.0556 75.8617L49.9999 99Z"
            fill="currentColor"
          />
          <path
            d="M100 49L76.8617 60.0556L76.8617 37.9444L100 49Z"
            fill="currentColor"
          />
          <path
            d="M1.57361e-06 49L23.1383 37.9444L23.1383 60.0556L1.57361e-06 49Z"
            fill="currentColor"
          />
        </g>
      </svg>
      <span
        className="DdgNodeContent--actionsItemText"
      >
        Set focus
      </span>
    </a>
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;

exports[`<DdgNodeContent> renders correctly when isFocalNode = true and focalNodeUrl = null 2`] = `
<div
  className="DdgNodeContent"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="DdgNodeContent--core is-focalNode"
    style={
      Object {
        "height": "100px",
        "transform": "translate(25px, 25px) scale(1.5)",
        "width": "100px",
      }
    }
  >
    <div
      className="DdgNodeContent--labelWrapper"
    >
      <h4
        className="DdgNodeContent--label"
        style={
          Object {
            "marginTop": "10px",
            "width": "20px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-service"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </h4>
      <div
        className="DdgNodeContent--label"
        style={
          Object {
            "paddingTop": "5px",
            "width": "30px",
          }
        }
      >
        <BreakableText
          className="BreakableText"
          text="some-operation"
          wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
        />
      </div>
    </div>
  </div>
  <div
    className="DdgNodeContent--actionsWrapper"
  >
    <a
      className="DdgNodeContent--actionsItem"
      onClick={[Function]}
      role="button"
    >
      <NewWindowIcon
        isLarge={false}
      />
      <span
        className="DdgNodeContent--actionsItemText"
      >
        View traces
      </span>
    </a>
  </div>
</div>
`;
